internal/fnruntime/runner.go

```diff
@@ -12,48 +12,100 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package pipeline provides struct definitions for Pipeline and utility
-// methods to read and write a pipeline resource.
-package cmdrender
+package fnruntime
 
 import (
 	"context"
+	goerrors "errors"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 
 	"github.com/GoogleContainerTools/kpt/internal/errors"
-	"github.com/GoogleContainerTools/kpt/internal/fnruntime"
+	"github.com/GoogleContainerTools/kpt/internal/printer"
 	"github.com/GoogleContainerTools/kpt/internal/types"
+	fnresult "github.com/GoogleContainerTools/kpt/pkg/api/fnresult/v1alpha2"
 	kptfilev1alpha2 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1alpha2"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
 	"sigs.k8s.io/kustomize/kyaml/fn/runtime/runtimeutil"
 	"sigs.k8s.io/kustomize/kyaml/kio"
 	"sigs.k8s.io/kustomize/kyaml/yaml"
 )
 
-// newFnRunner returns a fnRunner from the image and configs of
-// this function.
-func newFnRunner(ctx context.Context, f *kptfilev1alpha2.Function, pkgPath types.UniquePath) (kio.Filter, error) {
+// NewFunctionRunner returns a kio.Filter given a specification of a function
+// and it's config.
+func NewFunctionRunner(ctx context.Context, f *kptfilev1alpha2.Function, pkgPath types.UniquePath, fnResults *fnresult.ResultList) (kio.Filter, error) {
 	config, err := newFnConfig(f, pkgPath)
 	if err != nil {
 		return nil, err
 	}
 
-	cfn := &fnruntime.ContainerFn{
+	cfn := &ContainerFn{
 		Path:  pkgPath,
 		Image: f.Image,
 		Ctx:   ctx,
 	}
 
-	cfnw := &fnruntime.ContainerFnWrapper{
-		Fn: cfn,
+	return &FunctionRunner{
+		ctx:             ctx,
+		containerRunner: cfn,
+		fnResults:       fnResults,
+		filter: &runtimeutil.FunctionFilter{
+			Run:            cfn.Run,
+			FunctionConfig: config,
+		},
+	}, nil
+}
+
+// FunctionRunner wraps FunctionFilter and implements kio.Filter interface.
+type FunctionRunner struct {
+	ctx             context.Context
+	fnResults       *fnresult.ResultList
+	containerRunner *ContainerFn
```

Comment:
We cannot use `ContainerFn` here because in `eval` it can be `ExecFn`. I think we don't need this field? Looks like what we need from it are `Image`/`ExecPath` and `Ctx`. `Ctx` can be fetched from runner itself. We can have a `Image`/`ExecPath` in the `FunctionRunner`.